Source/CarthageKit/VersionFile.swift

```diff
@@ -0,0 +1,272 @@
+//
+//  VersionFile.swift
+//  Carthage
+//
+//  Created by Jason Boyle on 8/11/16.
+//  Copyright © 2016 Carthage. All rights reserved.
+//
+
+import Foundation
+import Runes
+import Argo
+import Curry
+import ReactiveSwift
+import ReactiveTask
+import Result
+import XCDBLD
+
+struct CachedFramework {
+	let name: String
+	let md5: String
+	
+	static let nameKey = "name"
+	static let md5Key = "md5"
+	
+	func toJSONObject() -> Any {
+		return [
+			CachedFramework.nameKey: name,
+			CachedFramework.md5Key: md5
+		]
+	}
+}
+
+extension CachedFramework: Decodable {
+	static func decode(_ j: JSON) -> Decoded<CachedFramework> {
+		return curry(self.init)
+			<^> j <| CachedFramework.nameKey
+			<*> j <| CachedFramework.md5Key
+	}
+}
+
+struct VersionFile {
+	let commitish: String
+	let xcodeVersion: String
+	
+	let macOS: [CachedFramework]?
+	let iOS: [CachedFramework]?
+	let watchOS: [CachedFramework]?
+	let tvOS: [CachedFramework]?
+	
+	static let commitishKey = "commitish"
+	static let xcodeVersionKey = "xcodeVersion"
+
+	subscript(_ platform: Platform) -> [CachedFramework]? {
+		switch platform {
+		case .macOS:
+			return macOS
+		case .iOS:
+			return iOS
+		case .watchOS:
+			return watchOS
+		case .tvOS:
+			return tvOS
+		}
+	}
+	
+	func toJSONObject() -> Any {
+		var dict: [String: Any] = [
+			VersionFile.commitishKey : commitish,
+			VersionFile.xcodeVersionKey : xcodeVersion
+		]
+		for platform in Platform.supportedPlatforms {
+			if let caches = self[platform] {
+				dict[platform.rawValue] = caches.map { $0.toJSONObject() }
+			}
+		}
+		return dict
+	}
+	
+	init(commitish: String, xcodeVersion: String, macOS: [CachedFramework]?, iOS: [CachedFramework]?, watchOS: [CachedFramework]?, tvOS: [CachedFramework]?) {
+		self.commitish = commitish
+		self.xcodeVersion = xcodeVersion
+		
+		self.macOS = macOS
+		self.iOS = iOS
+		self.watchOS = watchOS
+		self.tvOS = tvOS
+	}
+	
+	init?(url: URL) {
+		guard FileManager.default.fileExists(atPath: url.path),
+			let jsonData = try? Data(contentsOf: url),
+			let json = try? JSONSerialization.jsonObject(with: jsonData, options: .allowFragments),
+			let versionFile: VersionFile = Argo.decode(json) else {
+			return nil
+		}
+		self = versionFile
+	}
+
+	func satisfies(platform: Platform, commitish: String, xcodeVersion: String, binariesDirectoryURL: URL) -> SignalProducer<Bool, CarthageError> {
```

Comment:
I'm wondering if we should split this into 2 methods:

1. Check whether the cache is valid for a platform, commitish, and Xcode version
2. Check whether the frameworks on disk match the cache

That might simplify some of the logic?